dogechain-lab/dogechain
# [network] Feature: support static node

# Description

this PR add static node feature, static nodes configured for normal node not remove on runtime. unlike bootnode, node connections to static nodes is keep alive, if the connection is disconnected, node will actively try to reconnect static node (management connections is not affected by this PR)

command change

1. command `dogechain genesis` add `--staticnode` Flag
2. genesis file add `staticnodes` field
3. command `dogechain peers add` add `--static` Flag, dynamically add static nodes

genesis file simple:

```
{
.......
    "bootnodes": [
        "/ip4/10.5.0.15/tcp/1478/p2p/16Uiu2HAm7oonKfext8YgFCAMNWZKuXP2dpvnamCrN2VxjqTS54y3"
    ],
    "staticnodes": [
        "/ip4/10.5.0.15/tcp/1478/p2p/16Uiu2HAm7oonKfext8YgFCAMNWZKuXP2dpvnamCrN2VxjqTS54y3"
    ]
}
```    

# Changes include


# Checklist


## Testing


### Manual tests

1. create static nodes genesis file (1 boot node, 4 validator node, validator node configured boot node as a static node)
2. start up dogechain node
3. stop rpc node
4. check validator node is try reconnect rpc node
